core/src/test/scala/org/apache/spark/MiniKDCHelper.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import java.io.File
+import java.net.BindException
+
+import scala.util.control.NonFatal
+
+import org.apache.hadoop.minikdc.MiniKdc
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils
+
+trait MiniKDCHelper extends Logging {
+
+  def startMiniKdc(): MiniKdc = {
+    var kdc: MiniKdc = null
+    val kdcConf = MiniKdc.createConf()
+    kdcConf.setProperty(MiniKdc.DEBUG, "true")
+    var bindException = false
+    var kdcDir: File = null
+    // The port for MiniKdc service gets selected in the constructor, but will be bound
+    // to it later in MiniKdc.start() -> MiniKdc.initKDCServer() -> KdcServer.start().
+    // In meantime, when some other service might capture the port during this progress, and
+    // cause BindException.
+    // This makes our tests which have dedicated JVMs and rely on MiniKDC being flaky
+    //
+    // https://issues.apache.org/jira/browse/HADOOP-12656 get fixed in Hadoop 2.8.0.
+    // The workaround here is to capture the exception and retry, since we are using Hadoop 2.7.4
+    // as default.
+    // https://issues.apache.org/jira/browse/SPARK-31631
+    var numRetries = 1
+    do {
+      try {
+        bindException = false
+        kdcDir = Utils.createTempDir()
+        kdc = new MiniKdc(kdcConf, kdcDir)
+        kdc.start()
+      } catch {
+        case be: BindException if numRetries == 3 =>
+          logError(s"Failed setting up MiniKDC. Tried $numRetries times.");
+          throw be
+        case be: BindException =>
+          try {
+            Utils.deleteRecursively(kdcDir)
```

Comment:
Don't we need to stop it for full clean-up?

Reply:
Yes, seems better to invoke `stop` to clean-up the directory service, etc